app/manifest/v3/_base.json

```diff
@@ -47,6 +47,7 @@
   ],
   "default_locale": "en",
   "description": "__MSG_appDescription__",
+  "host_permissions": ["file://*/*", "http://*/*", "https://*/*"],
```

Comment:
I'm guessing that host permissions were needed for the `inpage.js` to establish a communication channel with the `contentscript.js`.

It would be interesting to find out whether `inpage.js` can communciate directly with the service worker though. It seems silly to have two content scripts, injected in two different ways.

Reply:
host_permissions were necessary to dynamically inject a content script into the "MAIN" world. Otherwise it'd stay isolated.

Reply:
> host_permissions were necessary to dynamically inject a content script into the "MAIN" world

Yep, host permissions were the missing piece I was stuck on for a while.

We *could* move all of our content scripts to `app-init` instead of the manifest, but since the manifest is the "right" way and this is sort of a "hack", I felt it made sense to do it the way I did.

Reply:
> We could move all of our content scripts to app-init instead of the manifest, but since the manifest is the "right" way and this is sort of a "hack", I felt it made sense to do it the way I did.

I think the point was not about where the scripts are defined, but whether we could entirely remove the contentscript.js running as an ISOLATED world script and forwarding between the MAIN world inpage.js and background. That'd be possible if a script in MAIN can talk to background worker. 

My best guess at this time is it might be possible using some chrome specific stuff that's explicitly not supported by browser-polyfill.

Reply:
I just tried this out locally, and it looks like this isn't an option. Chrome is treating the attempt as a normal 'external connection'. And since `externally_connectable` still doesn't support domain wildcards, there is no way for us to accept connections this way from any website. We're stuck with `contentscript.js` acting as a message relay for now.